File: scripts/parameters/cosmologicalParametersMonteCarlo.pl
#!/usr/bin/env perl
use strict;
use warnings;
use Cwd;
use lib $ENV{'GALACTICUS_EXEC_PATH'}."/perl";
use XML::Simple;
use PDL;
use PDL::NiceSlice;
use PDL::LinearAlgebra;
use PDL::MatrixOps;
use UNIVERSAL qw(isa);

# Generate sets of cosmological parameters drawn at random from the WMAP-9 constraints using the full covariance matrix.
# Andrew Benson (15-September-2010)

# Read the parameters and their covariances.
my $xml = new XML::Simple;
my $data = $xml->XMLin($ENV{'GALACTICUS_DATA_PATH'}."/static/cosmology/Cosmological_Parameters_WMAP-9.xml");
my $parameterCount = 0;
my %parameterMap;
foreach my $parameter ( @{$data->{'parameter'}} ) {
    $parameterMap{$parameter->{'label'}} = $parameterCount;
    ++$parameterCount;
}

# Create the covariance matrix and means vector.
my $mean       = pdl zeroes($parameterCount);
my $covariance = pdl zeroes($parameterCount,$parameterCount);
foreach my $parameterA ( @{$data->{'parameter'}} ) {
    my $indexA = $parameterMap{$parameterA->{'label'}};
    $mean(($indexA)) .= $parameterA->{'mean'};
    my @parametersB;
    if (isa($parameterA->{'parameter'},'ARRAY')) {
	@parametersB = @{$parameterA->{'parameter'}};
    } else {
	@parametersB = ( $parameterA->{'parameter'} );
    }
    foreach my $parameterB ( @parametersB ) {
	my $indexB = $parameterMap{$parameterB->{'label'}};
	$covariance(($indexA),($indexB)) .= $parameterB->{'covariance'};
	$covariance(($indexB),($indexA)) .= $parameterB->{'covariance'};
    }
}

# Perform a Cholesky decomposition on the covariance matrix.
my $choleskyDecomposed = mchol($covariance);

# Generate Gaussian random numbers.
my $deviates = grandom($parameterCount);

# Generate a set of parameters.
my $parameters = $mean + ($deviates x $choleskyDecomposed);

# Compute required parameters.
my $Omega_M            = $parameters(($parameterMap{'omega_M'}))/($parameters(($parameterMap{'H_0'}))/100.0)**2;
my $Omega_DE           = 1.0-$Omega_M;
my $Omega_b            = $parameters(($parameterMap{'omega_b'}))/($parameters(($parameterMap{'H_0'}))/100.0)**2;
my $sigma_8            = $parameters(($parameterMap{'sigma_8'}));
my $H_0                = $parameters(($parameterMap{'H_0'}));
my $powerSpectrumIndex = $parameters(($parameterMap{'n_s'}));

# Construct data for XML output.
my $parameterData =
{
    cosmologyParameters =>
    {
	value => "simple",
	"Omega_M"            => {value => $Omega_M           ->sclr()},
	"Omega_DE"           => {value => $Omega_DE          ->sclr()},
	"Omega_b"            => {value => $Omega_b           ->sclr()},
	"sigma_8"            => {value => $sigma_8           ->sclr()},
	"H_0"                => {value => $H_0               ->sclr()},
	"powerSpectrumIndex" => {value => $powerSpectrumIndex->sclr()}
    }
};

# Output data as XML.
my $xmlOutput = new XML::Simple (RootName=>"parameters");
print $xmlOutput->XMLout($parameterData);

exit;

File: scripts/parameters/cosmologicalParametersMonteCarlo.py
#!/usr/bin/env python3
# Generate sets of cosmological parameters drawn at random from the WMAP-9 constraints using the full covariance matrix.
# Andrew Benson (15-September-2010)

import os
import sys
import numpy as np
import xml.etree.ElementTree as ET
from xml.dom import minidom

# Read the parameters and their covariances.
dataPath = os.environ.get('GALACTICUS_DATA_PATH', '')
xmlFile = os.path.join(dataPath, 'static', 'cosmology', 'Cosmological_Parameters_WMAP-9.xml')
tree = ET.parse(xmlFile)
root = tree.getroot()

parameters = root.findall('parameter')
parameterCount = len(parameters)
parameterMap = {p.find('label').text: i for i, p in enumerate(parameters)}

# Create the covariance matrix and means vector.
mean       = np.zeros(parameterCount)
covariance = np.zeros((parameterCount, parameterCount))

for paramA in parameters:
    labelA = paramA.find('label').text
    indexA = parameterMap[labelA]
    mean[indexA] = float(paramA.find('mean').text)
    subParams = paramA.findall('parameter')
    for paramB in subParams:
        labelB = paramB.find('label').text
        indexB = parameterMap[labelB]
        cov = float(paramB.find('covariance').text)
        covariance[indexA, indexB] = cov
        covariance[indexB, indexA] = cov

# Perform a Cholesky decomposition on the covariance matrix.
choleskyDecomposed = np.linalg.cholesky(covariance)

# Generate Gaussian random numbers.
deviates = np.random.normal(size=parameterCount)

# Generate a set of parameters.
params = mean + choleskyDecomposed @ deviates

# Compute required parameters.
h = params[parameterMap['H_0']] / 100.0
Omega_M            = params[parameterMap['omega_M']] / h**2
Omega_DE           = 1.0 - Omega_M
Omega_b            = params[parameterMap['omega_B']] / h**2
sigma_8            = params[parameterMap['sigma_8']]
H_0                = params[parameterMap['H_0']]
powerSpectrumIndex = params[parameterMap['n_s']]

# Build XML output.
root_out = ET.Element('parameters')
cosmo = ET.SubElement(root_out, 'cosmologyParameters')
cosmo.set('value', 'simple')
for name, value in [
    ('Omega_M'           , Omega_M           ),
    ('Omega_DE'          , Omega_DE          ),
    ('Omega_b'           , Omega_b           ),
    ('sigma_8'           , sigma_8           ),
    ('H_0'               , H_0              ),
    ('powerSpectrumIndex', powerSpectrumIndex),
]:
    elem = ET.SubElement(cosmo, name)
    elem.set('value', str(value))

xmlStr = minidom.parseString(ET.tostring(root_out, encoding='unicode')).toprettyxml(indent='  ')
print(xmlStr)
